wcコマンド課題

Ryooo-k opened the pull request.

Ryooo-k committed:
<初回提出＞
※処理ベタ書き
※リファクタリング前
1.必要要件（行、単語数、バイト数）の出力処理実装
2.オプション（w,l,c）実装
3.複数ファイル指定可
4.複数ファイル指定時、合計値を出力

Ryooo-k committed:
Merge branch 'main' into wc

Ryooo-k committed:
＜標準入力時処理追加、リファクタリング＞
1.処理追加：ファイル指定がない場合、標準入力から読み取る
2.メソッド分け：標準入力・ファイル指定別に属性を読み込む
3.メソッド分割：属性値とtotalの取得を分割
4.メソッド削除：count_xxxのメソッドを削除
5.定数設定：出力幅設定の為、WIDTH定数を設定
6.メソッド追加：optionsを分岐とした属性削除メソッドの追加
7.その他：変数名・メソッド名など変更

yuuu commented on 05.wc/wc.rb:
```diff
@@ -0,0 +1,110 @@
+def main
+  options = parse_options
+  is_stdin = ARGV.empty?
+  loaded_attributes = is_stdin ? load_attributes_of_stdin : load_attributes_of_files
```
それぞれのメソッドをまとめていただくことでここは不要になると思いますが、仮に今のメソッドを変えない場合は以下のようにまとめたほうが読みやすいですね。

```ruby
output_text = if is_stdin
                to_output_text_from_stdin_attributes(load_attributes_of_stdin, options)
              else
                to_output_text_from_files_attributes(load_attributes_of_files, options)
              end
```

Ryo-ta090972 replied:
アドバイスありがとうございます🙇‍♂️

yuuu commented on 05.wc/wc.rb:
```diff
@@ -0,0 +1,110 @@
+def format_attributes(nested_attributes, width)
+  nested_attributes.map do |attributes|
+    attributes.values.map do |attribute|
+      attribute.instance_of?(Integer) ? "#{attribute.to_s.rjust(width)} " : attribute
```
出力したい属性が増えた場合や、数値であるものの `rjust` が不要な属性が出てきた場合に破綻する可能性があります。 カウント結果と `name` は分けて管理したほうが良いですね。

yuuu replied:
`attributes.values.map` を `attributes.map` にして、キー名で判定するのも手ですね。

yuuu commented on 05.wc/wc.rb:
```diff
@@ -0,0 +1,110 @@
+  end
+end
+
+def format_attributes(nested_attributes, width)
```
`nested_attributes` が何を指しているのか分かりづらいですね。すごく抽象化するなら `files`のような名前が考えられそうです。`file_counts` のような名前でも良いかもしれません。( `name` が含まれるので少し矛盾するような気もしますが )

Ryo-ta090972 replied:
命名にものすごく悩みました💦
アドバイスいただいた命名も含めて再度考えてみます！

Ryo-ta090972 commented on 05.wc/wc.rb:
```diff
@@ -0,0 +1,110 @@
+  end
+end
+
+def to_output_text_from_stdin_attributes(stdin_attributes, options)
```
ご指摘ありがとうございます！
一つにまとめてみます。

Ryo-ta090972 commented on 05.wc/wc.rb:
```diff
@@ -0,0 +1,110 @@
+  options
+end
+
+def load_attributes_of_stdin
```
承知しました！
修正いたします。

Ryooo-k committed:
1.メソッド変更：load_attributes_of_xxx⇒load_attributes
2.File処理修正：File.new(file_path)⇒File.open do 〜 end
3.メソッド変更：to_output_text_from_xxx⇒to_output_text
4.変数名変更：nested_attributes⇒files
5.処理変更：format_attributesメソッドの処理変更

Ryooo-k committed:
6.メソッド追加：calculate_attributes

yuuu commented on 05.wc/wc.rb:
```diff
@@ -0,0 +1,107 @@
+end
+
+def calculate_width_for_rjust(files, options)
+  if ARGV.empty?
```
ここは標準入力かどうかでロジックを分ける必要があるのでしょうか？単純に `files` で渡されたものから一番大きいものを選ぶだけだと何か不都合がありますか？

Ryo-ta090972 replied:
ご返信いただいた通り、 files で渡されたものから一番大きいものを選ぶ仕様に変更します。

yuuu commented on 05.wc/wc.rb:
```diff
@@ -0,0 +1,107 @@
+def calculate_max_int_length_whole_attributes(files)
+  files.map do |attributes|
+    attributes.values.map do |attribute|
+      attribute.to_s.length if attribute.instance_of?(Integer)
```
[こちら](https://github.com/Ryo-ta090972/ruby-practices/pull/9#discussion_r1398503437) のコメントはここにも該当するので対処をお願いいたします。

Ryo-ta090972 replied:
失念しておりました。。。
変更いたします。

yuuu commented on 05.wc/wc.rb:
```diff
@@ -0,0 +1,107 @@
+  copy_files_attributes = files_attributes.dup
+  copy_files_attributes << build_total_attributes(copy_files_attributes) if multiple_files?(copy_files_attributes)
+  width_for_rjust = calculate_width_for_rjust(copy_files_attributes, options)
+  output_files_attributes = options.all? { |_, bool| bool } ? copy_files_attributes : delete_attribute_by_option(copy_files_attributes, options)
```
`options.all?` が `true` の場合であっても `delete_attribute_by_option` を呼んで問題ないのではないでしょうか？(何も消されないだけなので)
そうすればここの三項演算子が無くせるのでシンプルになります。

Ryo-ta090972 replied:
`parse_options`メソッドを変更し、ご指摘箇所の三項演算子を無くすよう修正します。

yuuu commented on 05.wc/wc.rb:
```diff
@@ -0,0 +1,107 @@
+  }
+end
+
+def calculate_width_for_rjust(files, options)
```
全体的にメソッド名が具体的過ぎるので、不要な単語を省略できないか検討してみてください。例えばこのメソッド名だと `for_rjust` はなくても特に支障はないと思います。
参考: https://bootcamp.fjord.jp/articles/78

Ryo-ta090972 replied:
承知しました。
再考します！

Ryooo-k committed:
1.具体的なメソッド名を抽象化して短縮
2.parse_optionsメソッドの処理内容を変更
3.定数WIDTH_xxxを削除
4.widthは最大文字列で設定
